st2common/st2common/rbac/syncer.py

```diff
@@ -267,3 +295,105 @@ def _sync_user_role_assignments(self, user_db, role_assignment_dbs, role_assignm
+        """
+        groups = list(set(groups))
+
+        extra = {'user_db': user_db, 'groups': groups}
+        LOG.info('Synchronizing remote role assignments for user "%s"' % (str(user_db)),
+                 extra=extra)
+
+        # 1. Retrieve group to role mappings for the provided groups
+        all_mapping_dbs = GroupToRoleMapping.query(group__in=groups)
+        enabled_mapping_dbs = [mapping_db for mapping_db in all_mapping_dbs if
+                               mapping_db.enabled]
+        disabled_mapping_dbs = [mapping_db for mapping_db in all_mapping_dbs if
+                                not mapping_db.enabled]
+
+        if not all_mapping_dbs:
+            # No mapping found, return early
+            LOG.debug('No group to role mappings found for user "%s"' % (str(user_db)),
+                      extra=extra)
+            return ([], [])
+
+        # 2. Remove all the existing remote role assignments
+        remote_assignment_dbs = UserRoleAssignment.query(user=user_db.name, is_remote=True)
+
+        existing_role_names = [assignment_db.role for assignment_db in remote_assignment_dbs]
+        existing_role_names = set(existing_role_names)
+        current_role_names = set([])
+
+        for mapping_db in all_mapping_dbs:
+            for role in mapping_db.roles:
+                current_role_names.add(role)
+
+        # A list of new role assignments which should be added to the database
+        new_role_names = current_role_names.difference(existing_role_names)
+
+        # A list of role assignments which need to be updated in the database
+        updated_role_names = existing_role_names.intersection(current_role_names)
+
+        # A list of role assignments which should be removed from the database
+        removed_role_names = (existing_role_names - new_role_names)
+
+        # Also remove any assignments for mappings which are disabled in the database
+        for mapping_db in disabled_mapping_dbs:
+            for role in mapping_db.roles:
+                removed_role_names.add(role)
+
+        LOG.debug('New role assignments: %r' % (new_role_names))
+        LOG.debug('Updated role assignments: %r' % (updated_role_names))
+        LOG.debug('Removed role assignments: %r' % (removed_role_names))
+
+        # Build a list of role assignments to delete
+        role_names_to_delete = updated_role_names.union(removed_role_names)
+        role_assignment_dbs_to_delete = [role_assignment_db for role_assignment_db
+                                         in remote_assignment_dbs
+                                         if role_assignment_db.role in role_names_to_delete]
+
+        UserRoleAssignment.query(user=user_db.name, role__in=role_names_to_delete,
+                                 is_remote=True).delete()
+
+        # 3. Create role assignments for all the current groups
+        created_assignments_dbs = []
+        for mapping_db in enabled_mapping_dbs:
```

Comment:
I follow the approach of the existing syncer, but I plan to refactor it the near future - https://github.com/StackStorm/st2/issues/3310

Keep in mind that the end result is the same, the current approach just creates a bit more object churn but will refactor that in the near future.